### PR TITLE
Remove all marketing get parameters to minimize the cache using built-in FPC

### LINKED ISSUE #39266: [Issue] Remove all marketing get parameters to minimize the cache
This issue is automatically created based on existing pull request: magento/magento2#39099: Remove all marketing get parameters to minimize the cache

---------

### Description (*)

Same logic when use varnish (https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/PageCache/etc/varnish7.vcl#L84) but for build-in fpc

### Fixed Issues (if relevant)

1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)

1. Clear cache: bin/magento cache:clean
2. Run command: curl -I --location-trusted 'https://magento2.test/gear/bags.html?test=111&utm_source=123&gclid=abc' | grep -i X-Magento-Cache-Debug, it return x-magento-cache-debug: MISS => correct
3. Run command: curl -I --location-trusted 'https://magento2.test/gear/bags.html?test=111&utm_source=123&gclid=abc' | grep -i X-Magento-Cache-Debug, it return x-magento-cache-debug: HIT => correct
4. Run command: curl -I --location-trusted 'https://magento2.test/gear/bags.html?test=111' | grep -i X-Magento-Cache-Debug => it should return x-magento-cache-debug: HIT 

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by rogerdz**
### Description (*)

Same logic when use varnish (https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/PageCache/etc/varnish7.vcl#L84) but for build-in fpc

### Fixed Issues (if relevant)

1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)

1. Clear cache: bin/magento cache:clean
2. Run command: curl -I --location-trusted 'https://magento2.test/gear/bags.html?test=111&utm_source=123&gclid=abc' | grep -i X-Magento-Cache-Debug, it return x-magento-cache-debug: MISS => correct
3. Run command: curl -I --location-trusted 'https://magento2.test/gear/bags.html?test=111&utm_source=123&gclid=abc' | grep -i X-Magento-Cache-Debug, it return x-magento-cache-debug: HIT => correct
4. Run command: curl -I --location-trusted 'https://magento2.test/gear/bags.html?test=111' | grep -i X-Magento-Cache-Debug => it should return x-magento-cache-debug: HIT 

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#39266: Remove all marketing get parameters to minimize the cache

### CODE REVIEW

**Comment by rogerdz** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/App/PageCache/Identifier.php [hunk 1]
```diff
@@ -13,6 +13,23 @@
...
+        '/&?cof\=[^&]+/',
+        '/&?siteurl\=[^&]+/',
+        '/&?zanpid\=[^&]+/',
+        '/&?origin\=[^&]+/',
+        '/&?fbclid\=[^&]+/',
+        '/&?mc_(.*?)\=[^&]+/',
+        '/&?utm_(.*?)\=[^&]+/',
+        '/&?_bta_(.*?)\=[^&]+/',
+    ];
+
```
Comment (Bashev) [inline lib/internal/Magento/Framework/App/PageCache/Identifier.php]: Probably will be good idea to use the same list as is in the varnish configuration: https://github.com/magento/magento2/blob/71432aebbc5211378cc3ebb4f90a6ad9ee08648e/app/code/Magento/PageCache/etc/varnish7.vcl#L84

Comment (rogerdz) [reply lib/internal/Magento/Framework/App/PageCache/Identifier.php]: Yes, I used same list as varnish config

**Comment by rogerdz** [comment]
@magento run all tests

**Comment by sprankhub** [comment]
Nice. Maybe you want to update the parameters based on https://github.com/magento/magento2/pull/39188/files#diff-40e00b332f8dd95fbf48a312b9b8cfeee0d4e70bd7528c0e794c84c9e00c113fR85.

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run Database Compare, Functional Tests B2B, Functional Tests CE

**Review by engcom-Hotel — APPROVED** [review]

**Comment by sprankhub** [comment]
Please mind that https://github.com/magento/magento2/pull/39188 is also community-picked and is more complete. Specifically, it includes the changes from this PR. Hence, I think the solution from https://github.com/magento/magento2/pull/39188 should be merged.

**Comment by hostep** [comment]
@sprankhub, I think you got confused and meant to comment on https://github.com/magento/magento2/pull/35228
This PR here adds the extra parameters to the non-Varnish based FPC solution from Magento. But I believe the list should be updated with the list from https://github.com/magento/magento2/pull/39188 so they match 100%.

Ideally we should have one place where the entire list is defined and can be used both by the Varnish VCL file generator and the non-Varnish based FPC solution.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
If I got right, you’re removing cache parameters on Magento (php) side. 
for that purpose, Varnish does way better job, since it works way better and uses resources more efficiently.
I don’t see how someone would use built in full page cache in production, so I don’t think we should update it

Example use case: 
you added a new service, let’s say Klaviyo, that adds a dynamic value for each client.
Marketing team has sent an email to all subscribers and they starting clicking on link that goes to a single product page.
What we have right now is - the dynamic get parameter will be removed  by Varnish, and after the first request - it will put the result in cache. all next requests will not get to Magento/php - so they will be delivered to clients waaaay faster and won’t create load to web servers, database and redis.
if we accept your changes - this would mean that this dynamic get parameter will be removed on php side, which will result in additional slower responses, additional load to web servers, database and redis.

Having all that in mind, I don’t think we can accept this pull request, but instead, we should actualize the list in Varnish config.

Please correct me if I’m wrong

**Comment by ihor-sviziev** [comment]
Sorry, looks like I got confused a bit.
But still, it’s not clear if it worth fixing built in full page cache. Does anyone using it in production? If no - why would we support the list in 2 places?

**Comment by hostep** [comment]
Yes, people use Magento shops without Varnish in production as Varnish requires a big amount of memory and certain shops have soo little traffic it's not worth paying the extra cash for extra resources. In that case the built-in FPC using filesystem as storage is more then adequate enough. Some of our clients use this sort of setup.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Could you please cover your changes with unit tests?

**Comment by rogerdz** [comment]
@magento run all tests

**Comment by rogerdz** [comment]
@magento run all tests

**Comment by rogerdz** [comment]
@ihor-sviziev I'm updated unit test + marketing parameters

> Could you please cover your changes with unit tests?

I updated  unit test + marketing parameters

File: lib/internal/Magento/Framework/App/PageCache/Identifier.php [hunk 2]
```diff
@@ -13,6 +22,32 @@
  */
 class Identifier implements IdentifierInterface
 {
+    /**
+     * Pattern detect marketing parameters
+     */
+    public const PATTERN_MARKETING_PARAMETERS = [
+        '/&?gad_source\=[^&]+/',
```
Comment (ilnytskyi) [inline lib/internal/Magento/Framework/App/PageCache/Identifier.php]: I would suggest to extract this list to method that can be later pluginized. Or use dedicated class where those patterns can be injected via DI. Or even better to allow editing this list via admin panel.

If merchant uses some local marketing tools that are not part of this list then - hardcoded `public const PATTERN_MARKETING_PARAMETERS` cannot be easily overridden to remove custom tracking params.

As someone in CR suggested this service can also be used to generate Varnish config.

File: app/code/Magento/PageCache/Model/App/Request/Http/IdentifierForSave.php [hunk 3]
```diff
@@ -38,9 +48,11 @@ public function __construct(
      */
     public function getValue()
     {
+        $pattern = Identifier::PATTERN_MARKETING_PARAMETERS;
```
Comment (ilnytskyi) [inline app/code/Magento/PageCache/Model/App/Request/Http/IdentifierForSave.php]: This line is not extendable to append custom patterns. Dedicated service or public method visible to plugins would work better.

File: app/code/Magento/PageCache/Model/App/Request/Http/IdentifierForSave.php [hunk 4]
```diff
@@ -1,16 +1,26 @@
...
+ * All Rights Reserved.
+ *
+ * NOTICE: All information contained herein is, and remains
+ * the property of Adobe and its suppliers, if any. The intellectual
+ * and technical concepts contained herein are proprietary to Adobe
+ * and its suppliers and are protected by all applicable intellectual
+ * property laws, including trade secret and copyright laws.
+ * Dissemination of this information or reproduction of this material
+ * is strictly forbidden unless prior written permission is obtained from
+ * Adobe.
```
Comment (hostep) [inline app/code/Magento/PageCache/Model/App/Request/Http/IdentifierForSave.php]: This copyright header is too big.
Also the year should be the year the file was **created**, not when it was last updated.

So in this case the copyright header should be:
```
 * Copyright 2023 Adobe
 * All Rights Reserved.
```

Nothing else is needed. Not sure where you got that other paragraph from?

Same remark goes for the other files.

Comment (Bashev) [reply app/code/Magento/PageCache/Model/App/Request/Http/IdentifierForSave.php]: It's better to be moved to separate file  COPYING.txt or LICENSE.txt for better opportunity for update the policies. 

It's takes years to remove @author tag (comments block) from files, what will happens if all files requires updated license ...

Comment (in-session) [reply app/code/Magento/PageCache/Model/App/Request/Http/IdentifierForSave.php:13]: Please do not use any year dates, just use the default magento header here:

```
<?php
/**
 * Copyright © Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */
```

Comment (rogerdz) [reply app/code/Magento/PageCache/Model/App/Request/Http/IdentifierForSave.php]: new rule from magento coding standard (https://github.com/magento/magento-coding-standard/blob/develop/Magento2Framework/Sniffs/Header/CopyrightValidation.php)
I'm  copy from core
Have 94 file same format 
image

Comment (in-session) [reply app/code/Magento/PageCache/Model/App/Request/Http/IdentifierForSave.php]: It couldn't have been made any longer *lol* but thanks, I hadn't thought of that either

Comment (hostep) [reply app/code/Magento/PageCache/Model/App/Request/Http/IdentifierForSave.php]: So I discussed this with somebody from Adobe last week on Slack, he said to me that the new requirement from the Adobe legal team (which came after a lot of internal discussions), should be:

```
Copyright [year file created] Adobe
All Rights Reserved.
```

I totally agree that it's silly to change these copyright headers yet again, it has been changing way too often in the last 8 years or so. And I would love if we could just put it in an external file and reference to that and never have to change it again in the future, but the Adobe legal team disagrees with this approach for some unknown reason.

I also argued on Slack that the check in coding standards is too lose, as it only checks for the first line and you can basically put whatever you like after the first line, they agreed and they will fix this in a new version of coding standards someday in the future so it's clearer what is required.

So for now, can you please update it to only those 2 lines and use git to figure out the year each file was created and use that year. Thanks :)

**Comment by engcom-Dash** [comment]
Hello @rogerdz,

Thanks for the contributions!

Could you please address following comments: https://github.com/magento/magento2/pull/39099#discussion_r1819089463 and https://github.com/magento/magento2/pull/39099#discussion_r1819095028? Additionally, please update the copyright tag as suggested [here](https://github.com/magento/magento2/pull/39099#discussion_r1823318296).

If you're unable to make these changes, please let me know, and I will take care of it.

Thanks again!

**Comment by engcom-Dash** [comment]
@magento run all tests

File: app/code/Magento/PageCache/Model/App/Request/Http/IdentifierForSave.php (hunk 3 above)
Comment (engcom-Dash) [reply app/code/Magento/PageCache/Model/App/Request/Http/IdentifierForSave.php]: @ilnytskyi, I moved it to separate method. Could you please review it?

Comment (ilnytskyi) [reply app/code/Magento/PageCache/Model/App/Request/Http/IdentifierForSave.php]: @engcom-Dash lgtm 

`getMarketingParameterPatterns` can be pluginized to add custom patterns, thanks

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @rogerdz @engcom-Dash,

The changes looks good to me. But please address the failed automated tests. 

Thanks

**Comment by engcom-Dash** [comment]
@magento run all tests

**Review by engcom-Hotel — APPROVED** [review]

**Comment by engcom-Bravo** [comment]
Hi @rogerdz,

Thanks for your Contribution!!.

We have verified the issue in Latest 2.4-develop Magento Instance and the issue is not reproducible.Kindly refer the screenshots.

Screenshot 2024-11-22 at 12 16 46



The last request also showing `Hit`.Could you please let us know if we are missing anything.

Thanks.

**Comment by ihor-sviziev** [comment]
@engcom-Bravo, are you testing on built-in FPC without Varnish?

**Comment by engcom-Bravo** [comment]
Hi @ihor-sviziev,

Thanks for your update!!.

We have tested with Full Page Cache(built in) without varnish.

Screenshot 2024-11-22 at 18 19 24


Thanks.

**Comment by rogerdz** [comment]
@engcom-Bravo last request must call url without marketing parameter

**Comment by engcom-Bravo** [comment]
@magento run all tests

**Comment by engcom-Bravo** [comment]
Hi @rogerdz,

Thanks for the collaboration & contribution!

### :heavy_check_mark: QA Passed


**Preconditions:**

- Install fresh Magento **`2.4-develop`**

**Steps to reproduce**

- Clear cache: bin/magento cache:clean
- Run command: curl -I --location-trusted 'https://magento2.test/gear/bags.html?test=111&utm_source=123&gclid=abc' | grep -i X-Magento-Cache-Debug, it return x-magento-cache-debug: MISS => correct
- Run command: curl -I --location-trusted 'https://magento2.test/gear/bags.html?test=111&utm_source=123&gclid=abc' | grep -i X-Magento-Cache-Debug, it return x-magento-cache-debug: HIT => correct
- Run command: curl -I --location-trusted 'https://magento2.test/gear/bags.html?test=111' | grep -i X-Magento-Cache-Debug => it should return x-magento-cache-debug: HIT



**Before:** :heavy_multiplication_x: 



Screenshot 2024-11-25 at 09 45 49




**After:** :heavy_check_mark: 

Screenshot 2024-11-25 at 09 56 33


Builds are failed. Hence, moving this PR to **Extended Testing**.

Thanks.

**Comment by engcom-Dash** [comment]
@magento run Integration Tests, Functional Tests CE, Functional Tests B2B

**Comment by engcom-Dash** [comment]
@magento run Integration Tests, Functional Tests CE, Functional Tests B2B

**Comment by engcom-Dash** [comment]
The Functional B2B and CE test failures seem flaky and some of these failures are known issues. They are neither part of PR nor failing because of PR changes. Hence moving it to Merge in Progress.

**Functional Tests B2B**
Run 1: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39099/1875e843e30894817894cbc522dd0261/Functional/allure-report-b2b/index.html#categories/3a3a8306986cbb01e5245b23e6e5b238/6e05a95598af89e8/
image

Run 2: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39099/66f491bf0b44bceef57107d51d8cdf55/Functional/allure-report-b2b/index.html#categories/3a3a8306986cbb01e5245b23e6e5b238/5d0fa8a0b7cc1e95/
image

Known issues:
MultiShippingWithCreationNewCustomerAndAddressesDuringCheckoutTest: [ACQE-7229](https://jira.corp.adobe.com/browse/ACQE-7229)
StorefrontReuseCouponCodeAfterOrderCanceledTest: [ACQE-7230](https://jira.corp.adobe.com/browse/ACQE-7230)


**Functional Tests CE**
Run 1: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39099/977f45eeeb8f1049aa14572e5d9f74fc/Functional/allure-report-ce/index.html#categories/6d7e37752c3cc83cb3d2063131f608bb/30a31853636a9131/
image

Run 2: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39099/5eff451448e049fcb61454043d729cc1/Functional/allure-report-ce/index.html#categories/6d7e37752c3cc83cb3d2063131f608bb/476c47e6a72b24e0/
image

Known issue: 
VerifyDefaultWYSIWYGToolbarOnProductTestCest: [ACQE-7227](https://jira.corp.adobe.com/browse/ACQE-7227)

**Pull request merged**